Ez copy data selections

elizzhao opened the pull request:
Fixes #723 
Note: does not deal with select all across pages yet

Behavior:
Select and choose copy data
-if direct hard conflicts, asks user if they'd like to delete the existing (and potentially anything that references the existing such as sample sets, etc.) and override with the selected
-if soft conflicts, asks user if they'd like to link the selected entities to the existing 

Particularly looking for ways to improve:
-ExportDataModal: I use a lot of state and there may be better ways to write pieces of the code
-Warning/Alert messaging to user

elizzhao committed:
modal up

elizzhao committed:
rawls copy not working but modal is

elizzhao committed:
working?

elizzhao committed:
working version

elizzhao committed:
surface copy conflicts

elizzhao committed:
more error work

elizzhao committed:
backend works to save as new entityType

elizzhao committed:
language

elizzhao committed:
language

elizzhao committed:
trying to link entities

elizzhao committed:
hard conflict pathway beginnings

elizzhao committed:
override and copy working

elizzhao committed:
hardconflicts override and copy completed

elizzhao committed:
delete correct entities

elizzhao committed:
hard conflicts path done

note: has not reached end because of soft conflicts

elizzhao committed:
softConflicts showing up

elizzhao committed:
soft conflicts working maybe

elizzhao committed:
Merge branch 'dev' into ez-copy-data-selections

elizzhao committed:
copy delete copy link working

elizzhao committed:
pretty floating buttons!

elizzhao committed:
comments added

elizzhao commented:
Note: this implementation is good enough for now, we will revisit it later when our desired update semantics are better understood.

elizzhao committed:
Brett suggestions 1

elizzhao committed:
floating button changes

panentheos commented on src/components/ExportDataModal.js:
```diff
@@ -0,0 +1,228 @@
+    const errors = validate(
+      { selectedWorkspaceId },
+      { selectedWorkspaceId: { presence: true } },
+      { prettify: v => (validate.prettify(v)) }
```
nit: You only have to specify this option if you're overriding the default prettify function somehow.

panentheos commented on src/components/ExportDataModal.js:
```diff
@@ -0,0 +1,228 @@
+      selectedWorkspaceId: undefined,
+      error: undefined,
+      copying: false,
+      selectedEntityType: props.selectedDataType
```
This doesn't have to be in state since it doesn't change, so you could just use the prop instead.

panentheos commented on src/components/ExportDataModal.js:
```diff
@@ -0,0 +1,228 @@
+import validate from 'validate.js'
+
+
+class infoTile extends Component {
```
This is a good abstraction. A few nits:
Since it's stateless, you could use the 'functional' component notation which is slightly more compact.
PropTypes are probably overkill for something simple and file-local like this.
Component names should generally be title-cased, e.g. `InfoTile`

panentheos commented on src/components/ExportDataModal.js:
```diff
@@ -0,0 +1,228 @@
+          onDismiss()
+      }
+    }
+    this.copy()
```
The logic is well de-duplicated, but right now `copy` and `doOverride` are co-recursive, and rely on the state change to terminate. Even if this works now, it's risky because there's no guarantee that the results of a `setState` will be visible during the current tick. What about combining everything into a single function with a flow like:
```js
async copy() {
  if (hardConflicts.length) {
    // delete entries
  }
  // copy entries
}
```
with a couple bail-out points for stopping early to report conflicts.

elizzhao committed:
updates 2

breilly2 commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -510,9 +513,18 @@ const EntitiesContent = ajaxCaller(class EntitiesContent extends Component {
           })
         ])
       ]),
+      !!selectedEntities.length && h(FloatingActionButton, {
```
In #930, I changed the type of `selectedEntities` from an array to an object. You should be able to change this to `_.isEmpty(selectedEntities)` now and it will work with either type.

breilly2 commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -525,6 +537,11 @@ const EntitiesContent = ajaxCaller(class EntitiesContent extends Component {
+      copyingEntities && h(ExportDataModal, {
+        onDismiss: () => this.setState({ copyingEntities: false }),
+        workspace,
+        selectedEntities, selectedDataType: entityKey, runningSubmissionsCount
```
Also needs to change due to #930. Will need to be `selectedEntities: _.keys(selectedEntities)`. Whether now or later depends on which PR merges first. If this merges first, I'll take care of the change as part of #930.

breilly2 commented on src/components/FloatingActionButton.js:
```diff
@@ -43,7 +45,7 @@ export default class FloatingActionButton extends Component {
           }
         }, label),
-        icon(iconShape, { size: 25, style: { stroke: 'white', strokeWidth: 2 } })
+        icon(iconShape, { size: 25, style: { stroke: 'white', strokeWidth: 0.25 } })
```
Was this intentional? I don't notice any difference on my screen.

elizzhao replied:
I initially had made it thinner when I was trying to use the three-dot card menu since the wide thickness looked weird with that icon. However, since we're doing the two buttons for now instead, I'll change it back since it looks fine both ways!

elizzhao committed:
updates 3

elizzhao committed:
Lint

panentheos reviewed (approved):
Looks good.

katevoss commented on src/components/ExportDataModal.js:
```diff
@@ -0,0 +1,208 @@
+
+    return h(Modal, {
+      onDismiss,
+      title: 'Copy to Workspace',
```
Suggestion: "Copy Data to Workspace"?

katevoss commented on src/components/ExportDataModal.js:
```diff
@@ -0,0 +1,208 @@
+      onDismiss,
+      title: 'Copy to Workspace',
+      okButton: buttonPrimary({
+        tooltip: (hardConflicts.length !==0) ? 'Override existing entities, are you sure?' : Utils.summarizeErrors(errors),
```
I would rephrase as "Are you sure you want to override existing data?"

katevoss commented on src/components/ExportDataModal.js:
```diff
@@ -0,0 +1,208 @@
+      }),
+      (hardConflicts.length !== 0) && InfoTile({
+        infoStyle: errorStyle, iconName: 'error-standard',
+        content: 'The following entries already exist in the selected workspace. Please select CANCEL to go back or COPY to override the existing entities. '
```
How about: "Some of the following data already exists in the selected workspace. Click CANCEL to go back or COPY to override the existing data."
Thoughts:
- I think we want to avoid using the word "entities" as it doesn't mean much to users. 
- If a user selects 4 pieces of data and 2 pieces are duplicates, they will see this message. Correct? In which case I'd rephrase to "Some of"
- IMO a user _selects_ data and _clicks_ COPY. It's a semantic note but I figured I'd put it in while I'm here :)

katevoss commented on src/components/ExportDataModal.js:
```diff
@@ -0,0 +1,208 @@
+      }),
+      moreToDelete && InfoTile({
+        infoStyle: warningStyle, iconName: 'warning-standard',
+        content: 'In order to override the selected data entries, the following entries that reference the original data must ALSO be deleted.'
```
To clarify the UX, if a user deletes a data entry that another data entry references, the ref will be deleted. Is that correct?
In which case I'd recommend simplifying the message: "To override the selected data entries, the following entries that reference the original data will also be deleted."

katevoss commented on src/components/ExportDataModal.js:
```diff
@@ -0,0 +1,208 @@
+      }),
+      (softConflicts.length !== 0) && InfoTile({
+        infoStyle: warningStyle, iconName: 'warning-standard',
+        content: 'The following is linked to entries which already exist in the selected workspace. You may link the following to the existing entities by clicking COPY.'
```
This feels like it's missing a word, like "The following data is linked...". Also again note use of entities, which I don't know what it means to users.

katevoss commented:
Sorry to be adding comments late, but I noticed a few things:
- When I select entities to copy and click "Copy", the popup modal gives a dropdown list of workspaces to copy the data to. This dropdown includes the current workspace. Shouldn't we only include workspaces aside from the current workspace?
- If a user tries to copy duplicate data I get the message "The following entries already exist in the selected workspace. Please select CANCEL to go back or COPY to override the existing entities." While that message is in the modal, I can change the workspace destination. But that doesn't clear the message. When I click "COPY" I don't know whether I am copying data to the first workspace I selected, or the one I changed to.
- Added suggestions inline for wording

elizzhao commented:
> * If a user tries to copy duplicate data I get the message "The following entries already exist in the selected workspace. Please select CANCEL to go back or COPY to override the existing entities." While that message is in the modal, I can change the workspace destination. But that doesn't clear the message. When I click "COPY" I don't know whether I am copying data to the first workspace I selected, or the one I changed to.

So the logic around here is pretty delicate (and something we definitely want to improve on in the future) and wouldn't handle selecting a new workspace while you've already started down one of the conflict paths. For now, the solution I'm implementing is just removing the workspace selector drop-down menu after the first "Copy" click. 

> * Added suggestions inline for wording

Added these in, thanks! They're a lot clearer now :)
